### DISCUSSION TITLE
How to document outputs of a JCLExec or JobExec command

### DISCUSSION THREAD

**Discussion by dennis-behm** [IBM/dbb-zappbuild#529] [Q&A]
A user recently reached out to me and asked **how they could document outputs that are produced by a JCLExec of a JobExec command**. 

**Background**

DBB provides a very simple API to generate records. `MVSExec` and `CopyToPDS` allow users to define the outputs. When the step is executed, DBB generates the record automatically. 

The below snippet defines the outputs of the link step for the Cobol language script:
https://github.com/IBM/dbb-zappbuild/blob/a719842f8b9738133e375eaade8497289c30fc05/languages/Cobol.groovy#L358

You can read more about the available Build Report Records in the [DBB Documentation](https://www.ibm.com/docs/en/dbb/2.0?topic=apis-build-report). 

It's worth to check out the [JAVA Doc for DBB Build Records](https://www.ibm.com/docs/en/dbb/2.0?topic=apis-build-report) as well. FYI - with DBB 2.0.2 a new Record has been introduced - the **UnixRecord** for any tasks producing outputs on Unix System Services. 

**Document outputs in JCLExec, JobExec** 

`JCLExec` or `JobExec` don't provide a `output` method to define the `outputs` attribute for their records (worth opening an [IDEA](https://ideas.ibm.com) for it!), therefore users have implemented a workaround.  The way, I have seen users who wanted to document outputs of JCLExec or JobExec have chosen one of the below options. 

**Option 1** After evaluating the return code of the JCL, users have implemented an MVSExec command using [IEFBR14](https://www.ibm.com/docs/en/zos-basic-skills?topic=utilities-iefbr14-utility-do-almost-nothing) (the do nothing program on z/OS), which they use to document the (known) outputs of JCL step including an `deployType` including a reference to the `buildFile` (the relative path of the file such as `MortgageApplication/cobol/epscmort.cbl`).

**Option 2** If their use case it a bit different, some have gone the route to use the `AnyTypeRecord` to define their own record type, they they then parse during the packaging process. zAppBuild uses the `AnyTypeRecord` to create a `DELETE_RECORD` that can then be processed by the packaging steps: https://github.com/IBM/dbb-zappbuild/blob/a719842f8b9738133e375eaade8497289c30fc05/utilities/BuildReportUtilities.groovy#L43-L44

**Comment by Bill-Pratt** [comment]
Is there a way to use AnyTypeRecord to add this structure to the BuildReport?       
"outputs": [
{
	"dataset": "AL73418.COMP.SYSADATA(RRMAS001)",
        "deployType": "SYSDATA"
}
],

**Reply by dennis-behm** [reply]
Unfortunately not.

Please check out the Java Doc for  [AnyTypeRecord](https://www.ibm.com/docs/api/v1/content/SS6T76_2.0.0/javadoc/com/ibm/dbb/build/report/records/AnyTypeRecord.html)

As indicated it is just key value pairs; values could be lists of strings, but your sample contains another structure under outputs.

zAppBuild generates a custom DELETE_RECORD, that looks like this:
```
 "file": "jcls\/DELETEME.jcl",
            "deletedBuildOutputs": [
                JENKINS.ZDAT.RETIRE.JCL(DELETEME),
                JENKINS.ZDAT.RETIRE.SOURCE(DELETEME)
            ],
            "id": "DELETE_RECORD_1",
            "type": "DELETE_RECORD"
        },
```
Please note the difference comma-separated list of strings  **vs** a list of own data structures.
